### PR TITLE
JW-79 Packet memory management

### PR DESCRIPTION
**Pull request by Attil**

### CODE REVIEW

File: windows/vrouter_mod.c [hunk 1]
```diff
@@ -759,6 +768,8 @@ SxExtStartNetBufferListsIngress(
 			Switch->NdisSwitchHandlers.AddNetBufferListDestination(Switch->NdisSwitchContext, extForwardedNbls, &newDestination);
 		}
 
+
```
Comment (mmasztalerczuk) [inline windows/vrouter_mod.c]: Unnecessary blank lines

File: windows/vrouter_mod.c [hunk 2]
```diff
@@ -823,7 +834,19 @@ SxExtStartCompleteNetBufferListsIngress(
 	DbgPrint("SxExtStartCompleteNetBufferListsIngress\r\n");
 	UNREFERENCED_PARAMETER(ExtensionContext);
 
-	SxLibCompleteNetBufferListsIngress(Switch,
-		NetBufferLists,
-		SendCompleteFlags);
+	if (NetBufferLists->SourceHandle == Switch->NdisFilterHandle)
+	{
+		DbgPrint("Completing injected NBL...\r\n");
+		SxLibCompletedInjectedNetBufferLists(Switch, 1);
```
Comment (mmasztalerczuk) [inline windows/vrouter_mod.c]: Hardcoded value?

File: windows/vrouter_mod.c [hunk 3]
```diff
@@ -728,8 +733,11 @@ SxExtStartNetBufferListsIngress(
 
 	if (extForwardedNbls != NULL)
 	{
-		PNDIS_SWITCH_FORWARDING_DESTINATION_ARRAY broadcastArray;
+		PNDIS_SWITCH_FORWARDING_DESTINATION_ARRAY broadcastArray = NULL;
+
+
 		Switch->NdisSwitchHandlers.GetNetBufferListDestinations(Switch->NdisSwitchContext, extForwardedNbls, &broadcastArray);
+
```
Comment (mmasztalerczuk) [inline windows/vrouter_mod.c]: Unnecessery blank lines

File: include/vr_packet.h [hunk 4]
```diff
@@ -259,7 +259,10 @@ struct vr_packet {
     unsigned char vp_ttl;
     unsigned char vp_queue;
     unsigned char vp_priority:4,
-                  vp_notused:4;
+		vp_notused:4;
```
Comment (mmasztalerczuk) [inline include/vr_packet.h]: Unnecessary change

File: windows/vr_host.c [hunk 5]
```diff
@@ -1,12 +1,32 @@
...
 /* Defined in windows/vrouter_mod.c */
 extern PSX_SWITCH_OBJECT SxSwitchObject;
+extern NDIS_HANDLE SxNBLPool;
 extern PNDIS_RW_LOCK_EX AsyncWorkRWLock;
 
+struct host_os * vrouter_get_host(void);
+
+#define vrouter_host (vrouter_get_host())
+
+/* TODO: Change to extern linkage when dp-core/vr_stats.c is ported. */
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: Create JW for that.

File: windows/vr_host.c [hunk 6]
```diff
@@ -22,23 +42,63 @@ struct scheduled_work_cb_data {
 
 NDIS_IO_WORKITEM_FUNCTION deferred_work_routine;
 
+static PNET_BUFFER_LIST
+create_nbl(unsigned int size)
+{
+	void* ptr = ExAllocatePoolWithTag(NonPagedPool, size, SxExtAllocationTag);
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: ExAllocatePoolWithTag can return NULL. I think that you should add error check here.

File: windows/vr_host.c [hunk 7]
```diff
@@ -22,23 +42,63 @@ struct scheduled_work_cb_data {
...
+	MDL* mdl = NdisAllocateMdl(SxSwitchObject->NdisFilterHandle, ptr, size);
+	mdl->Next = NULL;
+	PNET_BUFFER_LIST nbl = NdisAllocateNetBufferAndNetBufferList(SxNBLPool, 0, 0, mdl, 0, size);
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: I think that you should add error check here, too.

File: windows/vr_host.c [hunk 8]
```diff
@@ -22,23 +42,63 @@ struct scheduled_work_cb_data {
...
+	_crt_va_end(args);
+
+	return printed;
+}
+static unsigned int win_get_cpu(void);
+
 static void *
 win_malloc(unsigned int size, unsigned int object)
 {
+	(object);
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: What is it?

File: windows/vr_host.c [hunk 9]
```diff
@@ -431,38 +678,48 @@ static int
...
+    UNREFERENCED_PARAMETER(size);
+
     size = 0;
     return NULL;
 }
 
 static void
 win_soft_reset(struct vrouter *router)
 {
+	(router);
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: Again, what is it?

File: windows/vr_host.c [hunk 10]
```diff
@@ -52,10 +112,13 @@ win_page_alloc(unsigned int size)
 }
 
 static void
-win_free(void *mem)
+win_free(void *mem, unsigned int object)
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: I don't understand this change. What change this new parameter?

File: windows/vr_host.c [hunk 11]
```diff
@@ -68,87 +131,224 @@ win_vtop(void *address)
 }
 
 static void
-win_page_free(void *mem)
+win_page_free(void *mem, unsigned int size)
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: Again, It looks unnecessary for me.

File: windows/vr_host.c [hunk 12]
```diff
@@ -68,87 +131,224 @@ win_vtop(void *address)
...
     if (mem)
         ExFreePoolWithTag(mem, SxExtAllocationTag);
 
     return;
 }
 
+void 
+win_assoc_packet_nb(PNET_BUFFER_LIST nbl, struct vr_packet* pkt)
+{
+	nbl->MiniportReserved[0] = pkt;
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: Why 0? Maybe add define with a meaningful name?

File: windows/vr_host.c [hunk 13]
```diff
@@ -68,87 +131,224 @@ win_vtop(void *address)
...
+    UNREFERENCED_PARAMETER(size);
+
     if (mem)
         ExFreePoolWithTag(mem, SxExtAllocationTag);
 
     return;
 }
 
+void 
+win_assoc_packet_nb(PNET_BUFFER_LIST nbl, struct vr_packet* pkt)
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: Where are you using this function?

File: windows/vr_host.c [hunk 14]
```diff
@@ -68,87 +131,224 @@ win_vtop(void *address)
...
+struct vr_packet*
+win_get_packet_from_nbl(PNET_BUFFER_LIST nbl)
+{
+	return nbl->MiniportReserved[0];
+}
+
+inline struct vr_packet *
+win_get_packet(PNET_BUFFER_LIST nbl, struct vr_interface *vif)
+{
+	struct vr_packet *pkt = ExAllocatePoolWithTag(NonPagedPool, sizeof(struct vr_packet), SxExtAllocationTag);
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: Add error check.

File: windows/vr_host.c [hunk 15]
```diff
@@ -68,87 +131,224 @@ win_vtop(void *address)
...
+	pkt->vp_head =
+		(unsigned char*)MmGetSystemAddressForMdlSafe(nb->CurrentMdl, LowPagePriority | MdlMappingNoExecute) + NET_BUFFER_CURRENT_MDL_OFFSET(nb);
+	if (!pkt->vp_head)
+		goto drop;
+	
+	pkt->vp_tail = pkt->vp_end = 0;
+
+	unsigned short length = (unsigned short) NET_BUFFER_DATA_LENGTH(nb);
+
+	pkt->vp_end = length;
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: It should be 0 or length? (because you set this on 0 in 171 line)

File: windows/vr_host.c [hunk 16]
```diff
@@ -68,87 +131,224 @@ win_vtop(void *address)
...
+
+	pkt->vp_len = 0;
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: I'm not sure, but it not should be length here?

File: windows/vr_host.c [hunk 17]
```diff
@@ -68,87 +131,224 @@ win_vtop(void *address)
...
+		nbl = pkt->vp_net_buffer_list;
+		if (!nbl)
+			return;
+		cpu = pkt->vp_cpu;
+	}
+	else {
+		cpu = win_get_cpu();
+	}
+
+	if (router)
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: router will be always null. It is correct?

File: windows/vr_host.c [hunk 18]
```diff
@@ -68,87 +131,224 @@ win_vtop(void *address)
...
+
+	return;
 }
 
 static struct vr_packet *
 win_palloc_head(struct vr_packet *pkt, unsigned int size)
 {
-    /* Dummy implementation */
-    return NULL;
+	PNET_BUFFER_LIST nbl = pkt->vp_net_buffer_list;
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: Are you sure that pkt can't be null? Error check?

File: include/vr_windows.h [hunk 19]
```diff
@@ -39,6 +39,10 @@ int vr_hash_nic(NDIS_IF_COUNTED_STRING string);
 struct vr_assoc* vr_get_assoc(NDIS_IF_COUNTED_STRING string);
 void vr_set_assoc_oid(NDIS_IF_COUNTED_STRING name, NDIS_SWITCH_PORT_ID port_id);
 
+struct host_os * vrouter_get_host(void);
```
Comment (sodar) [inline include/vr_windows.h]: Maybe use windows callbacks directly (not through `host_os` struct)?

File: windows/vrouter_mod.c [hunk 20]
```diff
@@ -823,7 +834,19 @@ SxExtStartCompleteNetBufferListsIngress(
...
+	else
+	{
+		DbgPrint("Completing non-injected NBL...\r\n");
+		SxLibCompleteNetBufferListsIngress(Switch,
+			NetBufferLists,
+			SendCompleteFlags);
+	}
+	struct vr_packet* pkt = NetBufferLists->MiniportReserved[0];
+	if (pkt != NULL)
+		windows_host.hos_pfree(pkt, 0);
```
Comment (sodar) [inline windows/vrouter_mod.c]: Does not build because of this line.

File: windows/vr_host.c (hunk 8 above)
Comment (michal-clapinski) [inline windows/vr_host.c]: Maybe you meant UNREFERENCED_PARAMETER(object); ?

File: windows/vr_host.c (hunk 10 above)
Comment (michal-clapinski) [inline windows/vr_host.c]: lh_page_free(void *address, unsigned int size)
This is from linux/vrouter_mod.c. Wouldn't it be better to use the same parameter names as in other platform implementations?

File: windows/vr_host.c (hunk 8 above)
Comment (michal-clapinski) [inline windows/vr_host.c]: You should probably use "UNREFERENCED_PARAMETER(object)" here.

File: windows/vr_host.c (hunk 5 above)
Comment (Attil) [inline windows/vr_host.c]: @sodar Could you please do that and include a description of the idea? This bit is from a commit authored by you.

File: windows/vr_host.c (hunk 10 above)
Comment (Attil) [inline windows/vr_host.c]: Please, do not confuse `lh_page_free` and `lh_free`. `win_free` is the equivalent of `lh_free`, which has the following signature: `static void lh_free(void *mem, unsigned int object)`.

File: windows/vr_host.c (hunk 11 above)
Comment (Attil) [inline windows/vr_host.c]: It's the equivalent of `static void lh_page_free(void *address, unsigned int size)`. By using the same parameter list as linux version, it's easier to maintain compatibility even if some parameters aren't referenced.

File: windows/vr_host.c (hunk 15 above)
Comment (Attil) [inline windows/vr_host.c]: Good catch, typo. L171 should be `vp_data` instead.

File: windows/vr_host.c (hunk 5 above)
Comment (sodar) [inline windows/vr_host.c]: I have created an issue in JIRA (JW-159).

File: windows/vr_host.c (hunk 16 above)
Comment (Attil) [inline windows/vr_host.c]: `skb_headlen` is used on Linux to calculate it, which return length of the head, which is 0 at the moment. I'm not completely sure it should be like that, but that's the way it looks like.

File: windows/vr_host.c (hunk 17 above)
Comment (Attil) [inline windows/vr_host.c]: Yes, please notice a comment at L212. Until v_router gets ported , it should be commented out. Then, after porting it and uncommenting everything will work.

**Comment by Attil** [comment]
I've addressed the issues. Please review again.

File: windows/vr_host.c (hunk 10 above)
Comment (michal-clapinski) [inline windows/vr_host.c]: Sorry, meant to write this somewhere else

File: windows/vr_host.c (hunk 11 above)
Comment (michal-clapinski) [inline windows/vr_host.c]: Exactly, also I think you should use the same names as in other systems' implementations, so win_page_free(void *address, unsigned int size)

File: windows/vrouter_mod.c [hunk 21]
```diff
@@ -823,7 +829,28 @@ SxExtStartCompleteNetBufferListsIngress(
...
+	else
+	{
+		DbgPrint("Completing non-injected NBL...\r\n");
+		SxLibCompleteNetBufferListsIngress(Switch,
+			NetBufferLists,
+			SendCompleteFlags);
+	}
+	struct vr_packet* pkt = NetBufferLists->MiniportReserved[0];
+	if (pkt != NULL)
+		vr_pfree(pkt, 0);
```
Comment (mmasztalerczuk) [inline windows/vrouter_mod.c]: Replace 0 with VP_DROP_DISCARD

File: windows/vrouter_mod.c [hunk 22]
```diff
@@ -823,7 +829,28 @@ SxExtStartCompleteNetBufferListsIngress(
...
+		SxLibCompletedInjectedNetBufferLists(Switch, count);
+	}
+	else
+	{
+		DbgPrint("Completing non-injected NBL...\r\n");
+		SxLibCompleteNetBufferListsIngress(Switch,
+			NetBufferLists,
+			SendCompleteFlags);
+	}
+	struct vr_packet* pkt = NetBufferLists->MiniportReserved[0];
```
Comment (mmasztalerczuk) [inline windows/vrouter_mod.c]: I'm not sure, but what is NetBufferLists is head of list and there is more elements (or NetBufferLists is null?)

File: windows/vrouter_mod.c [hunk 23]
```diff
@@ -115,6 +117,8 @@ SxExtCreateSwitch(
 	SxSwitchObject = Switch;
 	AsyncWorkRWLock = NdisAllocateRWLock(Switch->NdisFilterHandle);
 
+	SxNBLPool = vrouter_generate_pool();
```
Comment (sodar) [inline windows/vrouter_mod.c]: Also `SxNBLPool` is not freed when switch is deleted.

**Comment by Attil** [comment]
Fixed memory leaks which happened in certain circumstances and corrected some formatting issues.

File: windows/vr_host.c [hunk 24]
```diff
@@ -22,23 +42,88 @@ struct scheduled_work_cb_data {
 
 NDIS_IO_WORKITEM_FUNCTION deferred_work_routine;
 
+static PNET_BUFFER_LIST
+create_nbl(unsigned int size)
+{
+    void* ptr = ExAllocatePoolWithTag(NonPagedPool, size, SxExtAllocationTag);
+
+    if (ptr == NULL)
+        return NULL;
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: I think that you can use here STATUS_INSUFFICIENT_RESOURCES

File: windows/vr_host.c [hunk 25]
```diff
@@ -22,23 +42,88 @@ struct scheduled_work_cb_data {
...
+    }
+
+    mdl->Next = NULL;
+    PNET_BUFFER_LIST nbl = NdisAllocateNetBufferAndNetBufferList(SxNBLPool, 0, 0, mdl, 0, size);
+
+    if (nbl == NULL)
+    {
+        NdisFreeMdl(mdl); // Don't call ExFreePoolWithTag after creating an MDL
+        return NULL;
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: return STATUS_INSUFFICIENT_RESOURCES ?

File: windows/vr_host.c [hunk 26]
```diff
@@ -68,87 +156,228 @@ win_vtop(void *address)
...
+}
+
 static struct vr_packet *
 win_palloc(unsigned int size)
 {
-    /* Dummy implementation */
-    return NULL;
+    PNET_BUFFER_LIST nbl = create_nbl(size);
+
+    return win_get_packet(nbl, NULL);
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: For nbl == null it will be null pointer dereference.

File: windows/vr_host.c [hunk 27]
```diff
@@ -68,87 +156,228 @@ win_vtop(void *address)
...
+    if (pkt == NULL)
+        return NULL;
+
+    PNET_BUFFER_LIST nbl = pkt->vp_net_buffer_list;
+    if (nbl == NULL)
+        return NULL;
+
+    PNET_BUFFER_LIST nb_head = create_nbl(size);
+
+    struct vr_packet* npkt = win_get_packet(nb_head, pkt->vp_if);
```
Comment (mmasztalerczuk) [inline windows/vr_host.c]: For nbl == null it will be null pointer dereference.

File: windows/vr_host.c (hunk 24 above)
Comment (Attil) [inline windows/vr_host.c]: STATUS_INSUFFICIENT_RESOURCES is a define, which can in special cases result in a valid memory address.

**Comment by Attil** [comment]
Bump.
